Make parts of the BlockNavigationList overridable using slots

adamziel opened the pull request:
## Description

First attempt at #20748

This PR makes it possible to plug into `BlockNavigationList` using slots.

## How has this been tested?

1. Apply this PR and https://github.com/WordPress/gutenberg/pull/22210

### On experimental nav-menus page:

1. Go to `/wp-admin/admin.php?page=gutenberg-navigation`
1. Edit the text from the navigation structure panel like on the gif below:

1. Try with both top-level and nested navigation items

### In posts editor:

1. Go to posts > add new
1. Add navigation block with some nested items
1. Try editing titles through the navigator in block inspector (slow! performance fix: #21973):
<img width="840" alt="Zrzut ekranu 2020-04-29 o 13 35 11" src="https://user-images.githubusercontent.com/205419/80591654-53ee4e80-8a1e-11ea-882e-774f5cc6700f.png">

1. Make sure to use rich formatting like pressing enter or using keyboard shortcuts (e.g. cmd+b on mac to bold selected text).

1. Try editing titles through the navigator in modal:
<img width="847" alt="Zrzut ekranu 2020-04-29 o 13 35 17" src="https://user-images.githubusercontent.com/205419/80591648-52248b00-8a1e-11ea-96d8-725b9b41bba1.png">

1. Confirm global block navigation remains non-editable and only allows block selection on click:

<img width="752" alt="Zrzut ekranu 2020-04-29 o 13 36 15" src="https://user-images.githubusercontent.com/205419/80591719-72544a00-8a1e-11ea-8c9b-152867159ab0.png">

## Screenshots
![2020-04-28 16-01-11 2020-04-28 16_01_34](https://user-images.githubusercontent.com/205419/80496483-8ee27a80-8969-11ea-835a-f84b445d807d.gif)

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

adamziel committed:
First stab at editable menu items in navigator

Use Slot/Fill instead of hardcoded components

Disable slots/fills in global editor block navigator

Grab clientId from context instead of requiring a prop

Improve readability of block-navigation/list.js

Use context to determine whether or not given BlockNavigation may be customized with slots

Restore default value for selectBlock in navigation-structure-panel.js

Remove obsolete onChange passed to NavigationStructurePanel

Use constant value for BlockNavigationContext

Use memo() in BlockStyles

Remove performance changes from BlockStyles

Generalize BlockNavigationListItem and remove the BlockNavigationListItemFill for now

Rename useBlockNavigationSlots to withBlockNavigationSlots

Restore RichText in the navigator

Sort out BlockNavigationListItem and BlockNavigationListItemWrapper

Rename BlockNavigationListItemWrapper to BlockNavigationBranch

adamziel commented:
What's still missing here is distinguishing between the contextual/experimental block navigator and the global one - we don't want to make the text editable in the global navigator.

talldan commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -40,30 +40,20 @@ export default function BlockNavigationList( {
-								) }
-							</Button>
-						</div>
+						<ListItemSlot blockId={ block.clientId }>
```
You may be able to avoid the need to pass the `clientId` here by using something like `BlockContext` internally in the `ListItemSlot`, this seems to have the `clientId`:
https://github.com/WordPress/gutenberg/blob/6c26cdf1f97307b478e37d731351c4b1125e902b/packages/block-editor/src/components/block-list/block-wrapper.js#L36-L53

Or there might be another way. I think all the other slots rely on showing only content for the selected block, whereas this use case is a bit different.

talldan commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -193,6 +215,11 @@ function NavigationLinkEdit( {
 					/>
 				</PanelBody>
 			</InspectorControls>
+			<BlockNavigationListItemFill blockId={ clientId }>
```
You may be able to avoid the need to pass the `clientId` here by using something like `BlockContext` internally in the `ListItemSlot`, this seems to have the `clientId`:
https://github.com/WordPress/gutenberg/blob/6c26cdf1f97307b478e37d731351c4b1125e902b/packages/block-editor/src/components/block-list/block-wrapper.js#L36-L53

Or there might be another way. I think all the other slots rely on showing only content for the selected block, whereas this use case is a bit different.

talldan commented:
> What's still missing here is distinguishing between the contextual/experimental block navigator and the global one - we don't want to make the text editable in the global navigator.

We also have the same issue with some of the block toolbar items that we'd ideally not display on the block toolbar in the nav menus page:
https://github.com/WordPress/gutenberg/issues/21310

One idea might be to use an editor setting (e.g. `__experimentalDisableNavigationBlockNavigatorEditing` or similar) and adding conditional rendering in the block edit function.

Another option might be applying some filters to the block when registering. That approach would work well for changing the block `supports` settings to remove alignment options.

adamziel commented:
@talldan Super helpful notes - it thank you so much! Please keep them coming :-)

> We also have the same issue with some of the block toolbar items that we'd ideally not display on the block toolbar in the nav menus page - #21310

I believe the problem here is different. While #21310 is definitely relevant, here we want to make sure that the "global" navigator in post editor keeps working the same way, while pretty much every other navigator is editable:

<img width="685" alt="Zrzut ekranu 2020-04-29 o 13 30 51" src="https://user-images.githubusercontent.com/205419/80591256-a8dd9500-8a1d-11ea-9fe7-02b210e0b0f3.png">

I used contexts for now as they seem the cleanest way out.

adamziel commented:
It's working, although a tad bit slow - I'm going to profile it as my next step. Any suggestions and feedback welcome!

adamziel commented:
The root cause of slowness is that every time block attribute changes, like label after every key press, we re-render block styles preview in the inspector. This is really slow! I'll look into some optimizations around these styles. Alternatively we could disable rich text editing in the inspector for now.

adamziel commented:
Performance fix available in https://github.com/WordPress/gutenberg/pull/21973 - in the meantime let's get this one reviewed :-)

adamziel commented:
The performance fix is now merged.

talldan commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -77,7 +80,9 @@ function HeaderToolbar( { onToggleInserter, isInserterOpen } ) {
 			<EditorHistoryRedo />
 			<TableOfContents hasOutlineItemsDisabled={ isTextModeEnabled } />
-			<BlockNavigationDropdown isDisabled={ isTextModeEnabled } />
+			<BlockNavigationContext.Provider value={ blockNavigationContext }>
+				<BlockNavigationDropdown isDisabled={ isTextModeEnabled } />
+			</BlockNavigationContext.Provider>
```
I think the provider could be useful, as there are a few places where props are passed through multiple layers internally in the block navigation components (though some of that is not very visible due to the recursive nature of the component.)

It feels like the `BlockNavigationContext.Provider` should be an internal detail though. Probably as part of a wrapper around the top level `BlockNavigationList`.

adamziel replied:
If the provider was an internal detail, only used in `block-navigation/index.js`, how would you control the value of `withBlockNavigationSlots` from the `HeaderToolbar`?

talldan replied:
An option would be to pass it as a prop to `BlockNavigationDropdown` and then pass that prop to the context provider. I know that defeats some of the purpose of context, but it still achieves the goal of using context to pass the setting down to the leaf navigation item.

talldan commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -193,6 +222,14 @@ function NavigationLinkEdit( {
+					wrapperComponent="div"
+					onClick={ noop }
+				>
+					{ editField }
```
It's cool that this works so well.

The only thing I noticed is that being able to format the text (using shortcuts) is a bit unusual in the navigator, though I'm not sure how that could be handled in any other way.

talldan commented on packages/block-editor/src/components/index.js:
```diff
@@ -15,7 +15,12 @@ export { default as BlockEdit, useBlockEditContext } from './block-edit';
+export {
+	default as __experimentalBlockNavigationList,
+	BlockNavigationContext as __experimentalBlockNavigationContext,
+	ListItemFill as __experimentalBlockNavigationListItemFill,
```
It's verbose, but I think this should probably be `BlockNavigationListItemFill` internally as well for consistency.

talldan commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -40,30 +46,16 @@ export default function BlockNavigationList( {
-								) }
-							</Button>
-						</div>
+						<ListItemWrapper
```
Naming wise, it stood out to me that this is `ListItemWrapper` and yet it doesn't wrap the list item (`<li>`), though I guess it wraps `BlockNavigationListItem`.

I personally feel like making the slot implementation internal to `BlockNavigationListItem` and then making that component represent the whole `<li>` might provide a better separation:
```jsx
<BlockNavigationListItem
    block={ block }
    isSelected={ block.clientId === selectedBlockClientId }
    // ... other props
 />
        { showNestedBlocks &&
		!! block.innerBlocks &&
		!! block.innerBlocks.length && (
			<BlockNavigationList
				blocks={ block.innerBlocks }
				selectedBlockClientId={
					selectedBlockClientId
				}
				selectBlock={ selectBlock }
				parentBlockClientId={ block.clientId }
				showAppender={ showAppender }
				showNestedBlocks
			/>
		) }
</BlockNavigationListItem>
```

Using a weird notation that I just made up, the hierarchy would be something like:

- `BlockNavigationList` renders many
  - `BlockNavigationListItem` renders
    - `useBlockNavigationSlots` ? `BlockNavigationItemSlot` : `BlockNavigationItem`
    - `BlockNavigationList` (as `children`)

talldan reviewed (commented):
This is looking pretty solid already in testing. 🎉

talldan commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -18,6 +19,8 @@ import {
 import { Button } from '@wordpress/components';
 import { plus } from '@wordpress/icons';
 
+const blockNavigationContext = { useBlockNavigationSlots: false };
```
I suppose one final comment is that `use...` is often used for hooks these days. It probably wouldn't get confusing given it's clearly a property, but something like `hasBlockNavigationSlots` might be the safest option.

adamziel commented:
@talldan I addressed some of your feedback and removed the Fill for now. Tomorrow I will explore the navigation edit popover from https://github.com/WordPress/gutenberg/issues/21727#issuecomment-623399154

adamziel commented:
The next step is to improve the naming of a few components and rebase.

talldan commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -2,23 +2,28 @@
+import { BlockListBlockContext } from '../block-list/block';
+
+export const BlockNavigationContext = createContext( {
+	withBlockNavigationSlots: true,
```
My feeling is that it might be better to default this to false, so that it only has to be specified as `true` in the nav menus page. Changing less code in the post editor would be a positive.

adamziel replied:
Do you think we should only restrict that to nav menus page? This only disables those slots in the "global" navigation here:

<img width="752" alt="80591719-72544a00-8a1e-11ea-8c9b-152867159ab0" src="https://user-images.githubusercontent.com/205419/81058451-bf2c9a80-8ece-11ea-94a8-5d49b3c743a8.png">

But they remain active for example in the block inspector. The original issue actually was based of the post editor: https://github.com/WordPress/gutenberg/issues/20748

talldan replied:
I'm ok with either. The navigation block is experimental, so I think it's fine to keep it active there.

I think we'll end up removing those other navigators within the block eventually anyway, this seems like a great way forwards that would make them irrelevant - https://github.com/WordPress/gutenberg/issues/22113.

talldan commented on packages/block-editor/src/components/block-navigation/list.js:
```diff
@@ -95,3 +86,43 @@ export default function BlockNavigationList( {
+	selectBlock: () => {},
+};
+
+const BlockNavigationBranch = ( { withSlot, children, ...props } ) => {
```
The code looks, it might be nice to move BlockNavigationBranch and the slot/fills into a separate file.

adamziel committed:
Move BlockNavigationBranch to a different file

adamziel committed:
Refactor BlockNavigationContext to be an internal detail

adamziel commented:
@talldan I addressed all your feedback. Note that this feature is explicitly enabled in two inspectors in the post editor.

adamziel committed:
Lint

talldan commented on packages/block-editor/src/components/block-navigation/dropdown.js:
```diff
@@ -53,7 +53,7 @@ function BlockNavigationDropdownToggle( { isEnabled, onToggle, isOpen } ) {
 }
 
-function BlockNavigationDropdown( { isDisabled } ) {
+function BlockNavigationDropdown( { isDisabled, withBlockNavigationSlots } ) {
```
I think it'd be good to make this `__experimentalWithBlockNavigationSlots`, as it's a publicly exported prop, and it's hard to say what the future is, this may become the default at some point.

talldan commented on packages/block-editor/src/components/block-navigation/list-item.js:
```diff
@@ -0,0 +1,58 @@
+	const blockType = getBlockType( block.name );
+
+	return (
+		<div className="block-editor-block-navigation__item">
```
I just noticed some of the class names don't seem to be right after moving the code around. The root element should have the class name `block-editor-block-navigation-list-item` to reflect the component name.

talldan commented on packages/block-editor/src/components/index.js:
```diff
@@ -16,7 +16,12 @@ export { default as BlockEdit, useBlockEditContext } from './block-edit';
-export { default as __experimentalBlockNavigationList } from './block-navigation/list';
+export {
+	default as __experimentalBlockNavigationList,
+	BlockNavigationContext as __experimentalBlockNavigationContext,
```
I think this no longer needs to be exported now that it's only used internally in the BlockNavigation components

talldan reviewed (approved):
Happy with this, though I noticed a few minor things I could've caught in an earlier review, so sorry about that @adamziel!

In terms of the user experience, I still have a question over how a user will go about selecting a block that's editable in the navigation. Potentially they could click the icon.

The main concern for me would be having blocks that are editable next to blocks that are not editable and how we present to the user that clicking one does something different to clicking the other.

Having said that, this is only active in an experimental block, so lets move forwards.

We may also want to do some accessibility testing on this, or maybe audit the whole thing at some point if we can get some of the other PRs merged too.

adamziel committed:
Rename all occurences of withBlockNavigationSlots to __experimentalWithBlockNavigationSlots

adamziel committed:
Remove export of BlockNavigationContext

adamziel committed:
Adjust class names

adamziel commented:
@talldan 

> Happy with this, though I noticed a few minor things I could've caught in an earlier review, so sorry about that @adamziel!

No worries at all! I just addressed all of them

> Having said that, this is only active in an experimental block, so lets move forwards.

Agreed, we'll get there. One option would be to just have it enabled on the experimental nav-menus screen. And even if we decide not to use RichText in the navigator in the end, these slots are going to be useful for the ellipsis menu.

> We may also want to do some accessibility testing on this, or maybe audit the whole thing at some point if we can get some of the other PRs merged too.

👍

talldan commented:
Looks like there are some failing e2e tests that are relevant.

adamziel committed:
Remove the fill from edit.js for now

adamziel committed:
Update class name in e2e tests

adamziel committed:
Link

adamziel commented:
@talldan the failures were about the updated class name. I extracted the actual fill out of this PR into https://github.com/WordPress/gutenberg/pull/22210 - so this only adds the Slot/Fill machinery. Let's deal with any other test failures in 22210 and unblock the ellipsis menu.
